src/node_credentials.cc

```diff
@@ -33,11 +34,32 @@ bool linux_at_secure = false;
 
 namespace credentials {
 
-// Look up environment variable unless running as setuid root.
+#if !defined(__CloudABI__) && !defined(_WIN32)
+// Returns true if the current process has effective capabilities and the
+// passed-in capability is in that set.
+bool HasCapability(cap_value_t capability) {
+  DCHECK(cap_valid(capability));
+
+  cap_t cap = cap_get_proc();
+  if (cap == nullptr) {
+    return false;
+  }
+  cap_flag_value_t cap_flag_value;
+  cap_get_flag(cap, capability, CAP_EFFECTIVE, &cap_flag_value);
+
+  cap_free(cap);
+  return cap_flag_value == 1;
+}
+#endif
+
+// Look up the environment variable and allow the lookup if the current
+// process has the capability CAP_NET_BIND_SERVICE set. If the current process
+// does not have any capabilities set and the process is running as setuid root
+// then lookup will not be allowed.
 bool SafeGetenv(const char* key, std::string* text, Environment* env) {
 #if !defined(__CloudABI__) && !defined(_WIN32)
-  if (per_process::linux_at_secure || getuid() != geteuid() ||
-      getgid() != getegid())
+  if (!HasCapability(CAP_NET_BIND_SERVICE) && (per_process::linux_at_secure ||
+      getuid() != geteuid() || getgid() != getegid()))
```

Comment:
So ... we don't want to check whether `CAP_NET_BIND_SERVICE` is present or not; in fact, it is the *only* capability for which we don't want to check

And I'm also not sure whether it's okay to exclude the uid/gid checks in that case -- it shouldn't be, since we're still running with the privileges of another user in that case

Reply:
> it is the _only_ capability for which we don't want to check

Sorry, that is absolutely true :disappointed:  I'll take a look at this later today. Thanks

Reply:
I've update this now, would you be able to take another look? Thanks!